dali/operators/reader/loader/webdataset_loader.cc

```diff
@@ -0,0 +1,417 @@
+  empty = std::vector<Tensor<CPUBackend>>(ext_.size());
+  for (auto& tensor : empty) {
+    tensor.reserve(tensor_init_bytes_);
+  }
+}
+
+inline std::string GetExtension(const std::string& filepath) {
+  const size_t base_name_pos = filepath.find_last_of('/') + 1;
+  const size_t dot_pos = filepath.find_first_of('.', base_name_pos);
+  return filepath.substr(dot_pos + 1);
+}
+
+inline Index WebdatasetLoader::GetCurrentSampleIndex() const {
+  return wds_shards_prefixsums_[current_wds_shard_index_] + current_sample_index_;
+}
+
+inline void WebdatasetLoader::SetDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                                             std::vector<char>& sample_was_set,
+                                             const std::string& extension,
+                                             const std::string& source_info,
+                                             std::shared_ptr<void> data, int64_t size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      auto dtype_info = TypeTable::GetTypeInfo(dtypes_[component_index]);
+      sample[component_index].SetMeta(meta);
+      sample[component_index].ShareData(
+          data, size, {size / static_cast<int64_t>(dtype_info.size())}, dtype_info);
+      sample_was_set[component_index] = true;
+    }
+  }
+}
+
+
+inline uint8_t* WebdatasetLoader::ShareDataPointer(std::vector<Tensor<CPUBackend>>& sample,
+                                                   std::vector<char>& sample_was_set,
+                                                   const std::string& extension,
+                                                   const std::string& source_info,
+                                                   int64_t size) const {
+  DALIMeta meta;
+  meta.SetSourceInfo(source_info);
+  meta.SetSkipSample(false);
+
+  uint8_t* shared_tensor_data = nullptr;
+  for (size_t component_index : ext_map_.at(extension)) {
+    if (!sample_was_set[component_index]) {
+      sample[component_index].SetMeta(meta);
+      auto dtype_info = TypeTable::GetTypeInfo(dtypes_[component_index]);
+      if (shared_tensor_data == nullptr) {
+        if (sample[component_index].shares_data()) {
+          sample[component_index].Reset();
+        }
+        sample[component_index].reserve(size);
+        sample[component_index].Resize({size / static_cast<int64_t>(dtype_info.size())},
+                                       dtype_info);
+        shared_tensor_data = reinterpret_cast<uint8_t*>(sample[component_index].raw_mutable_data());
+      } else {
+        DALI_ENFORCE(size % static_cast<int64_t>(dtype_info.size()) == 0,
+                     "Index file at " + configs_[current_wds_shard_index_] +
+                         " reporting component sizes different to actual");
+        sample[component_index].ShareData(
```

Comment:
Why keep raw `shared_tensor_data`? Maybe it is better to keep the index and do:
```
sample[component_index].ShareData(&sample[shared_component_index]);
```

Reply:
alright, didn't know that was an option

Reply:
on second thought we can't since the shared data might have a different type